### PR TITLE
CA-178645: serialize SM plugin registration

### PR DESCRIPTION
**Pull request by zli**
This original issue only manifested in trunk-btrfs-3. RobertB suggested to upstream via trunk-ring3 as it won't do any harm here and we'll eventually need it.

### CODE REVIEW

File: ocaml/xapi/xapi_sm.ml [hunk 1]
```diff
@@ -92,22 +93,29 @@ let unregister_plugin ~__context query_result =
...
+				if rc.API.sM_type = driver then
+					try
+						info "Unregistering SM plugin %s (version %s)" driver query_result.version;
+						Db.SM.destroy ~__context ~self:rf
+					with e ->
+						warn "Ignore unregistering SM plugin failure: %s" (Printexc.to_string e))
+			(Db.SM.get_all_records ~__context)
+
+(* Lock to serialize SM plugin registration. Not necessary for unregister_plugin,
+   which is never called externally but only as a step in register_plugin. *)
```
Comment (thomassa) [inline ocaml/xapi/xapi_sm.ml]: At present `unregister_plugin` is not called externally from anywhere, but I feel we should ensure that this remains true in future. To do that we could have `xapi_sm.mli` without that function, or we could remove the function and move its `List.iter...` into `register_plugin`, replacing the call to `unregister_plugin`.

**Comment by thomassa** [comment]
Looks fine to me except the concern about protecting `unregister_plugin` against being called from elsewhere in future.

**Comment by zli** [comment]
Updated wrt review comments.

**Comment by thomassa** [comment]
@xen-git Retest this please

**Comment by thomassa** [comment]
I like it.
Something like the `_serialize_reg` function (with a different name) ought to be part of the standard `Mutex` module.
Merging.

**Pull request merged**